src/mongo/db/pipeline/expression.cpp

```diff
@@ -1729,9 +1727,17 @@ intrusive_ptr<ExpressionObject> ExpressionObject::parse(
 }
 
 intrusive_ptr<Expression> ExpressionObject::optimize() {
+    bool allValuesConstant = true;
     for (auto&& pair : _expressions) {
+        if (!dynamic_cast<ExpressionConstant*>(pair.second.get())) {
+            allValuesConstant = false;
+        }
         pair.second = pair.second->optimize();
```

Comment:
I think this line needs to go above the check of whether it is a constant value in order for your tests to work.